# data cleaning (np_address)

In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)

In [2]:
np_address = pd.read_csv("~/Downloads/aidatalab/Georgia_NPs_AddressesNPIs_new.csv")
np_address

,NPI,Entity_Type_Code,Provider_Last_Name,Provider_First_Name,Provider_Credentials,Street1,Street2,City,State,ZIP,Phone,Provider_Enumeration_Date,Provider_Sex,Provider_Taxonomy_Code,Provider_License_Number_1,Healthcare_Provider_Primary_Tax,Certification_Date,MEDICARE_SPECIALTY_CODE,MEDICARE_PROVIDER_SUPPLIER_TYPE,PROVIDER_TAXONOMY_DESCRIPTION_,NP_Type
0,1003095977,1,NICHOLAS,TRACY,NP,1968 PEACHTREE RD NW,PIEDMONT HO,ATLANTA,GA,303091281,4.046055e+09,10/29/2007,F,363LA2100X,185022,Y,NaN,50,Nurse Practitioner,Physician Assistants & Advanced Practice Nursi...,Acute Care NP
1,1003104670,1,DODD,JOHN,NP,2501 N PATTERSON ST,NaN,VALDOSTA,GA,316021735,2.292594e+09,7/18/2011,M,363LA2100X,RN117801,Y,NaN,50,Nurse Practitioner,Physician Assistants & Advanced Practice Nursi...,Acute Care NP
2,1003128240,1,SARNO,JENNIFER,ACNP-BC,780 CANTON RD NE,SUITE 400,MARIETTA,GA,300607241,7.704224e+09,7/3/2010,F,363LA2100X,177164,Y,NaN,50,Nurse Practitioner,Physician Assistants & Advanced Practice Nursi...,Acute Care NP
3,1003222191,1,SAUCEDO,ISRAEL,"APRN, FN",417 W 3RD AVE,NaN,ALBANY,GA,317011943,2.293121e+09,7/9/2014,M,363LA2100X,RN170526,N,6/29/2022,50,Nurse Practitioner,Physician Assistants & Advanced Practice Nursi...,Acute Care NP
4,1003280975,1,SHUMAKER-KIRK,CHARVETTE,NP-C,600 CELEBRATE LIFE PKWY,NaN,NEWNAN,GA,302658001,7.704007e+09,11/16/2015,F,363LA2100X,RN193221,Y,5/30/2023,50,Nurse Practitioner,Physician Assistants & Advanced Practice Nursi...,Acute Care NP
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14462,1982088340,1,STONE,SARAH,APRN-CNP,3278 MITCHELL BLVD,NaN,MOODY AFB,GA,316995529,2.292572e+09,7/16/2015,F,363LW0102X,RN176410,Y,NaN,50,Nurse Practitioner,Physician Assistants & Advanced Practice Nursi...,Womens Health NP
14463,1982210860,1,VINES,KRISTI,WHNP,10692 MEDLOCK BRIDGE RD,NaN,JOHNS CREEK,GA,300971890,4.044462e+09,9/16/2020,F,363LW0102X,RN255722,Y,9/16/2020,50,Nurse Practitioner,Physician Assistants & Advanced Practice Nursi...,Womens Health NP
14464,1982679189,1,KINNEY,KIMBERLY,"RN, NP",5354 REYNOLDS ST STE 422,NaN,SAVANNAH,GA,314056011,9.127219e+09,2/21/2006,F,363LW0102X,24164385,N,NaN,50,Nurse Practitioner,Physician Assistants & Advanced Practice Nursi...,Womens Health NP
14465,1992235469,1,CHAPMAN,NIJAH,WHNP,2665 N DECATUR RD STE 255,NaN,DECATUR,GA,300336176,4.702262e+09,6/15/2017,F,363LW0102X,RN228375,Y,NaN,50,Nurse Practitioner,Physician Assistants & Advanced Practice Nursi...,Womens Health NP


In [3]:
#they're all the same for every row so we can get rid of this col
# np_address['Entity_Type_Code'].nunique()
np_address['MEDICARE_PROVIDER_SUPPLIER_TYPE'].nunique()

1

In [4]:
#renames columns and drops unnessecary ones
np_address = np_address.rename(columns={'Provider_Last_Name': 'Last_Name'})
np_address = np_address.rename(columns={'Provider_First_Name': 'First_Name'})
np_address = np_address.rename(columns={'Provider_Credentials': 'Credentials'})
np_address = np_address.rename(columns={'Provider_Enumeration_Date': 'Enumeration_Date'})
np_address = np_address.rename(columns={'Provider_Sex': 'Sex'})
np_address = np_address.rename(columns={'Provider_Taxonomy_Code': 'Taxonomy_Code'})
np_address = np_address.rename(columns={'Provider_License_Number_1': 'License_Number'})
np_address = np_address.rename(columns={'Healthcare_Provider_Primary_Tax': 'Primary_Tax'})
np_address = np_address.rename(columns={'PROVIDER_TAXONOMY_DESCRIPTION_': 'Taxonomy_Description'})
np_address = np_address.rename(columns={'MEDICARE_PROVIDER_SUPPLIER_TYPE': 'Medicare_Supplier_Type'})
np_address = np_address.rename(columns={'MEDICARE_SPECIALTY_CODE': 'Medicare_Specialty_Code'})

np_address = np_address[['NPI', 'Last_Name', 'First_Name', 'Sex', 'Street1', 'Street2', 'City', 'ZIP', 'Credentials', 'Certification_Date', 'Enumeration_Date', 'License_Number', 'NP_Type', 'Taxonomy_Code', 'Taxonomy_Description', 'Medicare_Supplier_Type', 'Primary_Tax', 'Medicare_Specialty_Code']]
np_address

,NPI,Last_Name,First_Name,Sex,Street1,Street2,City,ZIP,Credentials,Certification_Date,Enumeration_Date,License_Number,NP_Type,Taxonomy_Code,Taxonomy_Description,Medicare_Supplier_Type,Primary_Tax,Medicare_Specialty_Code
0,1003095977,NICHOLAS,TRACY,F,1968 PEACHTREE RD NW,PIEDMONT HO,ATLANTA,303091281,NP,NaN,10/29/2007,185022,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
1,1003104670,DODD,JOHN,M,2501 N PATTERSON ST,NaN,VALDOSTA,316021735,NP,NaN,7/18/2011,RN117801,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
2,1003128240,SARNO,JENNIFER,F,780 CANTON RD NE,SUITE 400,MARIETTA,300607241,ACNP-BC,NaN,7/3/2010,177164,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
3,1003222191,SAUCEDO,ISRAEL,M,417 W 3RD AVE,NaN,ALBANY,317011943,"APRN, FN",6/29/2022,7/9/2014,RN170526,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50
4,1003280975,SHUMAKER-KIRK,CHARVETTE,F,600 CELEBRATE LIFE PKWY,NaN,NEWNAN,302658001,NP-C,5/30/2023,11/16/2015,RN193221,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14462,1982088340,STONE,SARAH,F,3278 MITCHELL BLVD,NaN,MOODY AFB,316995529,APRN-CNP,NaN,7/16/2015,RN176410,Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
14463,1982210860,VINES,KRISTI,F,10692 MEDLOCK BRIDGE RD,NaN,JOHNS CREEK,300971890,WHNP,9/16/2020,9/16/2020,RN255722,Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
14464,1982679189,KINNEY,KIMBERLY,F,5354 REYNOLDS ST STE 422,NaN,SAVANNAH,314056011,"RN, NP",NaN,2/21/2006,24164385,Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50
14465,1992235469,CHAPMAN,NIJAH,F,2665 N DECATUR RD STE 255,NaN,DECATUR,300336176,WHNP,NaN,6/15/2017,RN228375,Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50


## grouping np types

In [5]:
print(np_address["NP_Type"].value_counts())

NP_Type
Family NP                      7562
NP, No Subspecialty Noted      2593
Adult/Gero NP                   995
Pediatrics NP                   789
Psych/Mental Health NP          768
Acute Care NP                   744
Womens Health NP                332
Primary Care NP                 308
Neonatal NP                      99
OBGYN NP                         92
Neonatal Critical Care NP        67
Critical Care NP                 54
Pediatrics Critical Care NP      34
Community Health NP              18
Occupational Health NP            6
School NP                         6
Name: count, dtype: int64


groupings:

Acute Care NP: 744
Adult/Gero NP: 995
Community Health/Occupational Health/School NP: 30
Critical Care NP: 54
Family/Primary Care NP: 7870
Neonatal NP: 99
Neonatal Critical Care NP: 67
OBGYN/Womens Health NP: 424
Pediatrics NP: 789
Pediatrics Critical Care NP: 34
Psych/Mental Health NP: 768

In [6]:
mapping = {
    "Acute Care NP": "Acute Care NP",
    "Adult/Gero NP": "Adult/Gero NP",
    "Community Health NP": "Community/Occupational/School NP",
    "Occupational Health NP": "Community/Occupational/School NP",
    "School NP": "Community/Occupational/School NP",
    "Critical Care NP": "Critical Care NP",
    "Family NP": "Family/Primary Care NP",
    "Primary Care NP": "Family/Primary Care NP",
    "Neonatal NP": "Neonatal NP",
    "Neonatal Critical Care NP": "Neonatal Critical Care NP",
    "OBGYN NP": "OBGYN/Womens Health NP",
    "Womens Health NP": "OBGYN/Womens Health NP",
    "Pediatrics NP": "Pediatrics NP",
    "Pediatrics Critical Care NP": "Pediatrics Critical Care NP",
    "Psych/Mental Health NP": "Psych/Mental Health NP",
    "NP, No Subspecialty Noted": "NP, No Subspecialty Noted"
}

np_address["NP_Type_Grouped"] = np_address["NP_Type"].map(mapping)
np_address = np_address[['NPI', 'Last_Name', 'First_Name', 'Sex', 'Street1', 'Street2', 'City', 'ZIP', 'Credentials', 'Certification_Date', 'Enumeration_Date', 'License_Number', 'NP_Type_Grouped', 'Taxonomy_Code', 'Taxonomy_Description', 'Medicare_Supplier_Type', 'Primary_Tax', 'Medicare_Specialty_Code']]
np_address

,NPI,Last_Name,First_Name,Sex,Street1,Street2,City,ZIP,Credentials,Certification_Date,Enumeration_Date,License_Number,NP_Type_Grouped,Taxonomy_Code,Taxonomy_Description,Medicare_Supplier_Type,Primary_Tax,Medicare_Specialty_Code
0,1003095977,NICHOLAS,TRACY,F,1968 PEACHTREE RD NW,PIEDMONT HO,ATLANTA,303091281,NP,NaN,10/29/2007,185022,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
1,1003104670,DODD,JOHN,M,2501 N PATTERSON ST,NaN,VALDOSTA,316021735,NP,NaN,7/18/2011,RN117801,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
2,1003128240,SARNO,JENNIFER,F,780 CANTON RD NE,SUITE 400,MARIETTA,300607241,ACNP-BC,NaN,7/3/2010,177164,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
3,1003222191,SAUCEDO,ISRAEL,M,417 W 3RD AVE,NaN,ALBANY,317011943,"APRN, FN",6/29/2022,7/9/2014,RN170526,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50
4,1003280975,SHUMAKER-KIRK,CHARVETTE,F,600 CELEBRATE LIFE PKWY,NaN,NEWNAN,302658001,NP-C,5/30/2023,11/16/2015,RN193221,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14462,1982088340,STONE,SARAH,F,3278 MITCHELL BLVD,NaN,MOODY AFB,316995529,APRN-CNP,NaN,7/16/2015,RN176410,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
14463,1982210860,VINES,KRISTI,F,10692 MEDLOCK BRIDGE RD,NaN,JOHNS CREEK,300971890,WHNP,9/16/2020,9/16/2020,RN255722,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
14464,1982679189,KINNEY,KIMBERLY,F,5354 REYNOLDS ST STE 422,NaN,SAVANNAH,314056011,"RN, NP",NaN,2/21/2006,24164385,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50
14465,1992235469,CHAPMAN,NIJAH,F,2665 N DECATUR RD STE 255,NaN,DECATUR,300336176,WHNP,NaN,6/15/2017,RN228375,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50


## webscraping

In [7]:
found = pd.read_csv("~/Downloads/aidatalab/websrappingversion1.csv")
found

,NPI,Last_Name,First_Name,Sex,Street1,Street2,City,ZIP,Credentials,Certification_Date,Enumeration_Date,License_Number,NP_Type_Grouped,Taxonomy_Code,Taxonomy_Description,Medicare_Supplier_Type,Specialty,
0,1386213122,PATTERSON,JENNIFER,F,3222 SHRINE RD STE A,NaN,BRUNSWICK,315204357,ANP,9/24/24,6/24/21,RN331453,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Adult/Gero NP,NaN
1,1386487890,MCCRANIE,ASHTON,F,2258 WRIGHTSBORO RD STE 300,NaN,AUGUSTA,309044788,FNP-C,8/20/24,6/13/24,GAA-NP002402,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Family/Primary Care NP,NaN
2,1386659928,NIX,KELLY,F,145 N MEDICAL PKWY,NaN,WOODSTOCK,30189,"FNP, BC",4/19/23,7/30/06,112048,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Family/Primary Care NP,NaN
3,1386801900,THOMPSON,JENNIFER,F,1502 W 3RD ST,NaN,JACKSON,302331979,FNP,NaN,5/21/08,RN142971,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Family/Primary Care NP,NaN
4,1386805885,DAVIS,SHELLY,F,612 LOVE AVENUE,NaN,TIFTON,31794,FNP-BC,NaN,6/18/08,RN089222,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Family/Primary Care NP,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
646,1881786028,OKEREKE,CHIOMA,F,2317 AUSTELL RD SW,NaN,MARIETTA,300084537,"DNP, NP-",NaN,9/28/06,RN113110NP,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Pediatrics NP,NaN
647,1891832556,MULLIS,LISA,F,601 S 8TH ST,ER DEPT,GRIFFIN,302244213,NP,NaN,1/31/07,RN155767,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Acute Care NP,NaN
648,1952855926,PATRICK,JESSICA,F,1200 MEMORIAL DR,NaN,DALTON,307202529,F.N.P.,NaN,8/8/16,RN209632,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Family/Primary Care NP,NaN
649,1962838060,HIGGINS,AMY,F,5652 VETERANS PARKWAY,NaN,COLUMBUS,31904,C.P.N.P.,NaN,9/20/13,RN167875,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Pediatrics NP,NaN


In [8]:
# Ensure NPI is string so matching works
found['NPI'] = found['NPI'].astype(str)
np_address['NPI'] = np_address['NPI'].astype(str)

# Create a mapping: NPI → specialty
specialty_map = found.set_index('NPI')['Specialty']

# Overwrite NP_Type_Grouped ONLY where an NPI exists in found
np_address['NP_Type_Grouped'] = np_address['NPI'].map(specialty_map).fillna(np_address['NP_Type_Grouped'])

/var/folders/7q/7r1pkdyx07g6795dl8b86cs80000gp/T/ipykernel_4082/187791020.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  np_address['NPI'] = np_address['NPI'].astype(str)
/var/folders/7q/7r1pkdyx07g6795dl8b86cs80000gp/T/ipykernel_4082/187791020.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  np_address['NP_Type_Grouped'] = np_address['NPI'].map(specialty_map).fillna(np_address['NP_Type_Grouped'])


In [9]:
np_address

,NPI,Last_Name,First_Name,Sex,Street1,Street2,City,ZIP,Credentials,Certification_Date,Enumeration_Date,License_Number,NP_Type_Grouped,Taxonomy_Code,Taxonomy_Description,Medicare_Supplier_Type,Primary_Tax,Medicare_Specialty_Code
0,1003095977,NICHOLAS,TRACY,F,1968 PEACHTREE RD NW,PIEDMONT HO,ATLANTA,303091281,NP,NaN,10/29/2007,185022,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
1,1003104670,DODD,JOHN,M,2501 N PATTERSON ST,NaN,VALDOSTA,316021735,NP,NaN,7/18/2011,RN117801,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
2,1003128240,SARNO,JENNIFER,F,780 CANTON RD NE,SUITE 400,MARIETTA,300607241,ACNP-BC,NaN,7/3/2010,177164,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
3,1003222191,SAUCEDO,ISRAEL,M,417 W 3RD AVE,NaN,ALBANY,317011943,"APRN, FN",6/29/2022,7/9/2014,RN170526,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50
4,1003280975,SHUMAKER-KIRK,CHARVETTE,F,600 CELEBRATE LIFE PKWY,NaN,NEWNAN,302658001,NP-C,5/30/2023,11/16/2015,RN193221,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14462,1982088340,STONE,SARAH,F,3278 MITCHELL BLVD,NaN,MOODY AFB,316995529,APRN-CNP,NaN,7/16/2015,RN176410,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
14463,1982210860,VINES,KRISTI,F,10692 MEDLOCK BRIDGE RD,NaN,JOHNS CREEK,300971890,WHNP,9/16/2020,9/16/2020,RN255722,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
14464,1982679189,KINNEY,KIMBERLY,F,5354 REYNOLDS ST STE 422,NaN,SAVANNAH,314056011,"RN, NP",NaN,2/21/2006,24164385,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50
14465,1992235469,CHAPMAN,NIJAH,F,2665 N DECATUR RD STE 255,NaN,DECATUR,300336176,WHNP,NaN,6/15/2017,RN228375,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50


## filling np no subspeciatly based on credential (more accurate)

In [10]:
# --- 1. Create mapping dictionary from your groups ---

specialty_map = {
    "Acute Care NP": [
        "ACNPC-AG", "ACNP", "NP (ACNP", "ACNP-BC", "A.C.N.P.", "ACNPC"
    ],
    "Adult/Gero NP": [
        "AGACNP-B", "AGNP-C", "AGNP", "GNP", "AGACNP", "ACNP-AG", "MSN, ANP", "ANP",
        "AGPCNP-B", "AGACNP,", "AGPCNP", "A-GNP", "AGPCNP-C", "APRN, AG", "ANP-BC,",
        "AGACNP-C", "ANP, CNS", "AGPC-BC"
    ],
    "Family/Primary Care NP": [
        "FNP-C", "FNP", "FNP-BC", "MSN, FNP", "FNP, B.C", "DNP, FNP", "FNP, PHD", "CFNP",
        "APRN, FN", "FNP, PNP", "FNP-L", "FNP C", "MSN-FNP,", "F.N.P.", "FNPC"
    ],
    "Neonatal NP": [
        "NNP-BC"
    ],
    "NP No Subspecialty Noted": [
        # Very large group, but we do NOT use this group for replacement
    ],
    "OBGYN/Womens Health NP": [
        "WHNP-BC", "WHNP", "FP/WHNP-", "CNM MSN", "CNM"
    ],
    "Pediatrics Critical Care NP": [
        "PNP-PC/A", "CPNP-AC"
    ],
    "Pediatrics NP": [
        "CPNP", "PNP", "DNP, CPN", "CPNP-PC", "C.P.N.P."
    ],
    "Psych/Mental Health NP": [
        "PMHNP-BC", "APRN, PM"
    ]
}

# --- 2. Reverse-map each credential to its specialty group ---

credential_to_specialty = {}

for specialty, creds in specialty_map.items():
    for c in creds:
        credential_to_specialty[c] = specialty


# --- 3. Function to match credential to specialty ---
def map_specialty_from_credential(credential):
    if pd.isna(credential):
        return None
    
    credential_str = str(credential).strip()

    # exact match first
    if credential_str in credential_to_specialty:
        return credential_to_specialty[credential_str]

    # fuzzy match: contains string (helps for trailing commas, etc.)
    for cred, spec in credential_to_specialty.items():
        if isinstance(cred, str) and cred in credential_str:
            return spec

    return None


# --- 4. Apply: only replace rows where np_type_grouped == "NP, No Subspecialty Noted" ---

mask = np_address["NP_Type_Grouped"] == "NP, No Subspecialty Noted"

np_address.loc[mask, "NP_Type_Grouped"] = (
    np_address.loc[mask, "Credentials"]
    .apply(map_specialty_from_credential)
    .fillna("NP, No Subspecialty Noted")  # keep original if no match found
)

np_address

,NPI,Last_Name,First_Name,Sex,Street1,Street2,City,ZIP,Credentials,Certification_Date,Enumeration_Date,License_Number,NP_Type_Grouped,Taxonomy_Code,Taxonomy_Description,Medicare_Supplier_Type,Primary_Tax,Medicare_Specialty_Code
0,1003095977,NICHOLAS,TRACY,F,1968 PEACHTREE RD NW,PIEDMONT HO,ATLANTA,303091281,NP,NaN,10/29/2007,185022,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
1,1003104670,DODD,JOHN,M,2501 N PATTERSON ST,NaN,VALDOSTA,316021735,NP,NaN,7/18/2011,RN117801,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
2,1003128240,SARNO,JENNIFER,F,780 CANTON RD NE,SUITE 400,MARIETTA,300607241,ACNP-BC,NaN,7/3/2010,177164,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
3,1003222191,SAUCEDO,ISRAEL,M,417 W 3RD AVE,NaN,ALBANY,317011943,"APRN, FN",6/29/2022,7/9/2014,RN170526,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50
4,1003280975,SHUMAKER-KIRK,CHARVETTE,F,600 CELEBRATE LIFE PKWY,NaN,NEWNAN,302658001,NP-C,5/30/2023,11/16/2015,RN193221,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14462,1982088340,STONE,SARAH,F,3278 MITCHELL BLVD,NaN,MOODY AFB,316995529,APRN-CNP,NaN,7/16/2015,RN176410,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
14463,1982210860,VINES,KRISTI,F,10692 MEDLOCK BRIDGE RD,NaN,JOHNS CREEK,300971890,WHNP,9/16/2020,9/16/2020,RN255722,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
14464,1982679189,KINNEY,KIMBERLY,F,5354 REYNOLDS ST STE 422,NaN,SAVANNAH,314056011,"RN, NP",NaN,2/21/2006,24164385,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50
14465,1992235469,CHAPMAN,NIJAH,F,2665 N DECATUR RD STE 255,NaN,DECATUR,300336176,WHNP,NaN,6/15/2017,RN228375,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50


## filling missing specialties

In [11]:
undefined = np_address[np_address['NP_Type_Grouped'] == 'NP, No Subspecialty Noted']
undefined

,NPI,Last_Name,First_Name,Sex,Street1,Street2,City,ZIP,Credentials,Certification_Date,Enumeration_Date,License_Number,NP_Type_Grouped,Taxonomy_Code,Taxonomy_Description,Medicare_Supplier_Type,Primary_Tax,Medicare_Specialty_Code
9539,1003022310,KRECHMERY,KAREN,F,11 UPPER RIVERDALE RD SW,NaN,RIVERDALE,302742615,NP,NaN,5/15/2007,146180,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
9540,1003027434,BROWN,MICHELLE,F,655 1ST ST,NaN,MACON,312012852,N.P.,NaN,5/25/2007,67038,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
9541,1003114604,HORN,LAUREN,F,97 CHURCH ST,NaN,TAYLORSVILLE,301781902,NP-C,NaN,3/5/2011,RN188946,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
9542,1003151028,BLACK,CASEY,F,125 EAGLE SPRING DR,NaN,STOCKBRIDGE,302816328,NP,NaN,12/3/2012,RN146749,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
9544,1003199480,WILDES,DEANNA,F,4750 WATERS AVENUE,SUITE 500,SAVANNAH,314046261,"RN, BSN",9/13/2022,9/22/2011,RN077963,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12124,1992427595,WILLIAMS,SHELBY,F,455 S MAIN ST STE 103,NaN,HINESVILLE,313134354,NP,9/12/2022,9/12/2022,RN270412,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
12125,1992514574,WHITLOCK,WHITNEY,F,777 HEMLOCK ST,NaN,MACON,312012102,NaN,2/17/2025,1/7/2025,209822,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
12128,1992732085,COOK,GAYE,F,5 EXECUTIVE CIR,NaN,SAVANNAH,314063345,NP,NaN,6/26/2006,RN063377,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
12130,1992764708,CARPENTER,FRANCINE,F,ONE FREEDOM WAY,CHARLIE NOR,AUGUSTA,309046285,N.P.,NaN,3/21/2006,RN053788,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50


In [12]:
#nps matched from izzie web scraping
2593-1942

651

In [13]:
#nps matched from izzie web scraping + matching on credentials
2593-1626

967

In [17]:
undefined.to_csv('NEW_missing_specialty.csv', index=False)

In [16]:
#how many people are missing credentials and np specialty entirely
miss_cred = undefined[undefined['Credentials'].isna()]
miss_cred

,NPI,Last_Name,First_Name,Sex,Street1,Street2,City,ZIP,Credentials,Certification_Date,Enumeration_Date,License_Number,NP_Type_Grouped,Taxonomy_Code,Taxonomy_Description,Medicare_Supplier_Type,Primary_Tax,Medicare_Specialty_Code
9573,1013312529,ONYIRIMBA,ESTHER,F,870 CRESTMARK DR,NaN,LITHIA SPRINGS,301222665,NaN,NaN,10/27/2014,RN205364,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
9608,1023576170,NEWMAN,BRANDI,F,902 N 7TH ST,NaN,CORDELE,310153234,NaN,3/3/2021,3/7/2019,140369,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
9611,1023619764,ONWUBIKO,CHINWENWA,F,300 COLONIAL CENTER PKWY,NaN,ROSWELL,300764899,NaN,11/6/2020,11/6/2020,RN225814,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
9615,1023733300,SMITH,ERIN,F,1336 HIGHWAY 54 W BLDG 500,NaN,FAYETTEVILLE,302144574,NaN,11/18/2022,10/10/2022,RN219283,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
9617,1023744935,ENGLISH,HEATHER,F,145 CATOOSA CIR,NaN,RINGGOLD,307368077,NaN,7/25/2022,7/25/2022,31551,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12095,1982239497,FU,YING,F,777 W PEACHTREE ST,NaN,NORCROSS,300711868,NaN,3/4/2020,3/4/2020,RN252505,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
12096,1982294625,BRADSHAW,KELLEY,F,1364 CLIFTON RD NE,NaN,ATLANTA,303221059,NaN,5/17/2021,1/26/2021,RN301684,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
12113,1992145155,TURNER,RACHEL,F,21 COMMERCE PKWY,NaN,ADAIRSVILLE,301032009,NaN,NaN,7/2/2013,RN202294,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
12115,1992180517,VEGA,JULIA,F,5670 PEACHTREE DUNWOODY RD STE 1100,NaN,ATLANTA,303424795,NaN,9/18/2020,7/24/2015,216789,"NP, No Subspecialty Noted",363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50


In [ ]:
# num nps with either no credentials or general credentials that have no specialty notes


In [14]:
#print(missing['Taxonomy_Code'].unique()) => only returns '363L00000X' which is the generic NP code
#but they have different credentials!
print(undefined['Credentials'].unique())

['NP' 'N.P.' 'NP-C' 'WHNP-BC' 'RN, BSN' 'FNP-C' 'PMHNP-BC' 'AGACNP-B'
 'CNP' 'FNP' 'APRN' 'ACNPC-AG' 'AGNP-C' 'APRN,BC' nan 'CPNP' 'FNP-BC'
 'DNP' 'MSN, FNP' 'APRN/NP' 'RN 03380' 'FNP, B.C' 'AGNP' 'NURSE PR'
 'DNP, FNP' 'DNP, APR' 'FNP, PHD' 'CFNP' 'A.R.N.P.' 'APN' 'GNP' 'AGACNP'
 'RN' 'RN 12088' 'ARNP' 'PHD, CAN' 'WHNP' 'EDD, RN,' 'MSN, APR' 'APRN, FN'
 'ACNP-AG' 'MSN, ANP' 'FNP, PNP' 'FNP-L' 'NP, AANP' 'PNP-PC/A' 'RN, MSN,'
 'ACNP' 'CNM/ARNP' 'BSN NP' 'DNP, MSN' 'FP/WHNP-' 'FNP C' 'ANP' 'RN, CPNP'
 'AGPCNP-B' 'MD' 'MSN,FNP-' 'N.P.C.' 'NPC' 'NP (ACNP' 'AGACNP,' 'RN 05389'
 'ACNP-BC' 'CPNP-AC' 'APRN, NP' 'MSN' 'PA' 'MSN, NP-' 'APRNC' 'AGPCNP'
 'PNP' 'MSN, OCN' 'ANP-BC' 'RN NP' 'APRN, BC' 'CRNP' 'A-GNP' 'AGPCNP-C'
 'R.N., M.' 'APRN GAA' 'C-NP' 'A.C.N.P.' 'MSN, RN,' 'MS, APRN' 'PH.D. AP'
 'DCNP' 'APRN, AG' 'ANP-BC,' 'ACNPC' 'AGACNP-C' 'NP, PHD' 'ANP, CNS'
 'DNP, CPN' 'MSN-FNP,' 'MS, NP-C' 'NP-C, DN' 'NP, DNP' 'APRN-C' 'C.R.N.P.'
 'RN-NP' 'N.P.-C' 'APRN-BC' 'DNP, NP-' 'F.N.P.' 'FNPC' 'CPN

In [10]:
#list the credentials that each np_type has in the np_address df and use that list of credentials to map 'NP, No Subspecialty Noted' nps to an NP_Type
defined = np_address[np_address['NP_Type_Grouped'] != 'NP, No Subspecialty Noted']
cred_by_type = (
    defined.groupby('NP_Type_Grouped')['Credentials']
    .unique()
    .apply(lambda x: sorted([str(i).strip().upper() for i in x if pd.notna(i)]))
)

for np_type, creds in cred_by_type.items():
    print(f"\n{np_type}:")
    print(creds)


Acute Care NP:
['A.C.N.P', 'ACAG DNP', 'ACAGNP', 'ACGNP-BC', 'ACNP', 'ACNP-AG', 'ACNP-BC', 'ACNP-C', 'ACNPC-AG', 'ADVANCED', 'AG ACNP', 'AG-ACNP', 'AG-ACNP-', 'AGAC-DNP', 'AGAC-NP', 'AGACNP', 'AGACNP,', 'AGACNP-B', 'AGACNP-C', 'AGACNP-D', 'AGANCP-B', 'AGCNP', 'AGNPC-AG', 'APN', 'APRN', 'APRN, AC', 'APRN, AG', 'APRN, BC', 'APRN, FN', 'APRN, NP', 'APRN,BC', 'APRN-BC', 'APRN-BC,', 'APRNBC', 'ARNP', 'C.N.P.', 'CNP', 'CPNP', 'CPNP-AC', 'CRNA', 'CRNP', 'DNP', 'DNP, AG-', 'DNP, AGA', 'DNP, APR', 'DNP, NP', 'DNP-AGAC', 'FNP', 'FNP-BC', 'FNP-C, A', 'MPH, AG-', 'MS, MSN,', 'MSN, ACN', 'MSN, AG-', 'MSN, AGA', 'MSN, APR', 'MSN, ARN', 'MSN, CEN', 'MSN, MSP', 'MSN, RN', 'MSN, RN,', 'N.P.', 'NP', 'NP, RN', 'NP-BC', 'NP-C', 'NPC', 'NURSE PR', 'PMHNP-BC', 'PNP-BC', 'R.N., MS', 'RN', 'RN MSN A', 'RN,  ACN', 'RN, ACNP', 'RN, MSN,', 'RN,MSN,A', 'RN218258']

Adult/Gero NP:
['A-GNP', 'A-GNP-C', 'A-GPCNP', 'A.N.P.-B', 'A.P.R.N.', 'A.R.N.P', 'AANP-C', 'ACCNS-AG', 'ACNP', 'ACNP-C', 'ACNPC-AG', 'ADULT GE', 'AD

In [11]:
#credential -> most common NP_Type_Grouped
cred_to_type = (
    defined.groupby('Credentials')['NP_Type_Grouped']
    .agg(lambda x: x.value_counts().idxmax())  # most frequent NP_Type for that credential
    .dropna()
)

#check the mapping
mapping_df = cred_to_type.reset_index().rename(columns={'index': 'Credentials', 'NP_Type_Grouped': 'Mapped_NP_Type'})
mapping_df

,Credentials,Mapped_NP_Type
0,A-GNP,Adult/Gero NP
1,A-GNP-C,Adult/Gero NP
2,A-GPCNP,Adult/Gero NP
3,A.C.N.P,Acute Care NP
4,A.N.P.-B,Adult/Gero NP
...,...,...
557,WHNP-BC.,Adult/Gero NP
558,WHNP-BP,OBGYN/Womens Health NP
559,WHPN,OBGYN/Womens Health NP
560,WNHP-BC,OBGYN/Womens Health NP


In [12]:
#may group generic NPs with family/primary since that makes up the large majority of NPs nationally
#verified by frequency!!!!!!
test = mapping_df[mapping_df['Credentials'].isin(['NP', 'N.P.'])]
test

,Credentials,Mapped_NP_Type
375,N.P.,Family/Primary Care NP
383,NP,Family/Primary Care NP


In [13]:
#fill in missing NP_Type_Grouped values
mask = np_address['NP_Type_Grouped'] == 'NP, No Subspecialty Noted'
np_address.loc[mask, 'NP_Type_Grouped'] = np_address.loc[mask, 'Credentials'].map(cred_to_type)
np_address

,NPI,Last_Name,First_Name,Sex,Street1,Street2,City,ZIP,Credentials,Certification_Date,Enumeration_Date,License_Number,NP_Type_Grouped,Taxonomy_Code,Taxonomy_Description,Medicare_Supplier_Type,Primary_Tax,Medicare_Specialty_Code
0,1003095977,NICHOLAS,TRACY,F,1968 PEACHTREE RD NW,PIEDMONT HO,ATLANTA,303091281,NP,NaN,10/29/2007,185022,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
1,1003104670,DODD,JOHN,M,2501 N PATTERSON ST,NaN,VALDOSTA,316021735,NP,NaN,7/18/2011,RN117801,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
2,1003128240,SARNO,JENNIFER,F,780 CANTON RD NE,SUITE 400,MARIETTA,300607241,ACNP-BC,NaN,7/3/2010,177164,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
3,1003222191,SAUCEDO,ISRAEL,M,417 W 3RD AVE,NaN,ALBANY,317011943,"APRN, FN",6/29/2022,7/9/2014,RN170526,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50
4,1003280975,SHUMAKER-KIRK,CHARVETTE,F,600 CELEBRATE LIFE PKWY,NaN,NEWNAN,302658001,NP-C,5/30/2023,11/16/2015,RN193221,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14462,1982088340,STONE,SARAH,F,3278 MITCHELL BLVD,NaN,MOODY AFB,316995529,APRN-CNP,NaN,7/16/2015,RN176410,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
14463,1982210860,VINES,KRISTI,F,10692 MEDLOCK BRIDGE RD,NaN,JOHNS CREEK,300971890,WHNP,9/16/2020,9/16/2020,RN255722,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
14464,1982679189,KINNEY,KIMBERLY,F,5354 REYNOLDS ST STE 422,NaN,SAVANNAH,314056011,"RN, NP",NaN,2/21/2006,24164385,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50
14465,1992235469,CHAPMAN,NIJAH,F,2665 N DECATUR RD STE 255,NaN,DECATUR,300336176,WHNP,NaN,6/15/2017,RN228375,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50


In [14]:
#verify how many were filled
filled_count = np_address.loc[mask, 'NP_Type_Grouped'].notna().sum()
print(f"Filled {filled_count} previously 'No Subspecialty' NPs")

Filled 2230 previously 'No Subspecialty' NPs


In [15]:
#missing 2593 - 2230 specialties = 363
not_grouped = np_address[np_address['NP_Type_Grouped'].isna()]
not_grouped

,NPI,Last_Name,First_Name,Sex,Street1,Street2,City,ZIP,Credentials,Certification_Date,Enumeration_Date,License_Number,NP_Type_Grouped,Taxonomy_Code,Taxonomy_Description,Medicare_Supplier_Type,Primary_Tax,Medicare_Specialty_Code
9573,1013312529,ONYIRIMBA,ESTHER,F,870 CRESTMARK DR,NaN,LITHIA SPRINGS,301222665,NaN,NaN,10/27/2014,RN205364,NaN,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
9588,1013790096,COOK,KENDRA,F,123 WEAVER RD STE B,NaN,BLAIRSVILLE,305123155,APRN/NP,8/15/2023,8/15/2023,RN236809,NaN,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
9597,1023144847,GUNN,JOYCE,F,811 HEMLOCK ST,NaN,MACON,312012144,RN 03380,NaN,2/26/2007,RN 033806 NP,NaN,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
9602,1023285319,MOYE,ANGIE,F,6602 WATERS AVE BLDG C,NaN,SAVANNAH,314062778,"FNP, B.C",NaN,5/15/2008,RN164257,NaN,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
9608,1023576170,NEWMAN,BRANDI,F,902 N 7TH ST,NaN,CORDELE,310153234,NaN,3/3/2021,3/7/2019,140369,NaN,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12095,1982239497,FU,YING,F,777 W PEACHTREE ST,NaN,NORCROSS,300711868,NaN,3/4/2020,3/4/2020,RN252505,NaN,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
12096,1982294625,BRADSHAW,KELLEY,F,1364 CLIFTON RD NE,NaN,ATLANTA,303221059,NaN,5/17/2021,1/26/2021,RN301684,NaN,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
12113,1992145155,TURNER,RACHEL,F,21 COMMERCE PKWY,NaN,ADAIRSVILLE,301032009,NaN,NaN,7/2/2013,RN202294,NaN,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
12115,1992180517,VEGA,JULIA,F,5670 PEACHTREE DUNWOODY RD STE 1100,NaN,ATLANTA,303424795,NaN,9/18/2020,7/24/2015,216789,NaN,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50


In [16]:
#which in not_grouped are NOT in miss_cred (should be 363-288=75)
weird = not_grouped[not_grouped['Credentials'].notna()]
weird

,NPI,Last_Name,First_Name,Sex,Street1,Street2,City,ZIP,Credentials,Certification_Date,Enumeration_Date,License_Number,NP_Type_Grouped,Taxonomy_Code,Taxonomy_Description,Medicare_Supplier_Type,Primary_Tax,Medicare_Specialty_Code
9588,1013790096,COOK,KENDRA,F,123 WEAVER RD STE B,NaN,BLAIRSVILLE,305123155,APRN/NP,8/15/2023,8/15/2023,RN236809,NaN,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
9597,1023144847,GUNN,JOYCE,F,811 HEMLOCK ST,NaN,MACON,312012144,RN 03380,NaN,2/26/2007,RN 033806 NP,NaN,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
9602,1023285319,MOYE,ANGIE,F,6602 WATERS AVE BLDG C,NaN,SAVANNAH,314062778,"FNP, B.C",NaN,5/15/2008,RN164257,NaN,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
9626,1033153077,STEMP,LINDA,F,6667 VERNON WOODS DR STE A19,NaN,SANDY SPRINGS,303283215,"FNP, PHD",2/23/2022,6/16/2006,RN170552,NaN,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50
9633,1033413877,LALANDE,SHERRIE,F,308 DEEP SOUTH FARM RD,NaN,BLAIRSVILLE,305122218,A.R.N.P.,8/14/2024,1/10/2011,5009949,NaN,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
11929,1922506906,BEDINGFIELD,MARY,F,980 WOODSTOCK PKWY STE 300,NaN,WOODSTOCK,301884871,"MSN,NP-C",NaN,1/25/2018,RN158380,NaN,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
11939,1922852581,CHEATEM,SHANTRICE,F,167 ARYOLA DR,NaN,BAINBRIDGE,398177660,AGPC-BC,5/16/2024,4/11/2024,APRN11032253,NaN,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
11955,1932582301,JIMENEZ,SARAH,F,1120 15TH ST,NaN,AUGUSTA,30912,"APN-C, A",NaN,7/3/2015,26NJ00573000,NaN,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50
11983,1942233762,GIBSON,OLIVIA,F,515 RIVERSIDE DR,NaN,WAYCROSS,315015316,RN/APRN,11/15/2023,7/7/2006,RN113495,NaN,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50


In [17]:
#these are ones with credentials that the other nurses didn't have so we have to assign them to a NP_Group_Type based on what these credentials typically mean
print(weird['Credentials'].unique())

['APRN/NP' 'RN 03380' 'FNP, B.C' 'FNP, PHD' 'A.R.N.P.' 'RN 12088'
 'PHD, CAN' 'EDD, RN,' 'FNP, PNP' 'FNP-L' 'NP, AANP' 'PNP-PC/A' 'CNM/ARNP'
 'BSN NP' 'DNP, MSN' 'FP/WHNP-' 'MD' 'N.P.C.' 'NP (ACNP' 'RN 05389'
 'APRNC' 'MSN, OCN' 'RN NP' 'APRN GAA' 'A.C.N.P.' 'PH.D. AP' 'DCNP'
 'ACNPC' 'NP, PHD' 'ANP, CNS' 'MSN-FNP,' 'MS, NP-C' 'NP-C, DN' 'RN-NP'
 'AOCNP' 'MSN, PHD' 'RN, CDE' 'ADULT NU' 'DNP, MBA' 'FNP-BC B' 'RN PNP'
 'RNFA' 'CRNFA, N' 'FNP-C, G' 'MS, ARNP' 'FNPC FAM' 'CNM MSN' 'APRN-NP'
 'F.N.P.-B' 'APRN AGA' 'GNP-BC' 'R.N.-N.P' 'C.N.P' 'CPT,FNP-' 'M.N, APR'
 'APRN WHN' 'FNB-BC' 'NP-ADVAN' 'APRN, CN' 'APRN., F' 'APRN OR' 'FNP-BC,N'
 'NP, RNC,' 'LPN' 'MSN,NP-C' 'AGPC-BC' 'APN-C, A' 'RN/APRN']


In [18]:
#repeat mapping procedure with new credentials
np_address.loc[np_address['NP_Type_Grouped'].isna() & np_address['Credentials'].notna() & 
    np_address['Credentials'].str.contains(
        'ACNP|NP-ADVAN|APRN OR|A.C.N.P.|CRNFA|DCNP|AOCNP|AGPC-BC|RNFA|NP \(ACNP', 
        case=False, na=False), 'NP_Type_Grouped'] = 'Acute Care NP'

np_address.loc[np_address['NP_Type_Grouped'].isna() & np_address['Credentials'].notna() & 
    np_address['Credentials'].str.contains(
        'APN-C, A|ANP|GNP|ADULT NU|APRN AGA|GNP-BC', 
        case=False, na=False), 'NP_Type_Grouped'] = 'Adult/Gero NP'

np_address.loc[np_address['NP_Type_Grouped'].isna() & np_address['Credentials'].notna() & 
    np_address['Credentials'].str.contains(
        'CNS|RN, CDE', 
        case=False, na=False), 'NP_Type_Grouped'] = 'Community/Occupational/School NP'

np_address.loc[np_address['NP_Type_Grouped'].isna() & np_address['Credentials'].notna() & 
    np_address['Credentials'].str.contains(
        'LPN|C\\.N\\.P|MSN, PHD|PH\\.D\\. AP|N\\.P\\.C\\.|FNP|FNPC|F\\.N\\.P\\.|FNB|MSN-FNP|FNP-L|NP, AANP|BSN NP|MS, NP-C|MSN,NP-C|FNP-BC|FNP-C|FNPC FAM|CPT,FNP-|F.N.P.-B|FNP-BC,N|FNP, PHD|FNP, PNP|FNP, B.C|MS, ARNP|MSN, OCN|NP-C, DN', 
        case=False, na=False), 'NP_Type_Grouped'] = 'Family/Primary Care NP'

np_address.loc[np_address['NP_Type_Grouped'].isna() & np_address['Credentials'].notna() & 
    np_address['Credentials'].str.contains(
        'PNP|NP, RNC|PNP-PC', 
        case=False, na=False), 'NP_Type_Grouped'] = 'Neonatal NP'

np_address.loc[np_address['NP_Type_Grouped'].isna() & np_address['Credentials'].notna() & 
    np_address['Credentials'].str.contains(
        'WHNP|CNM|FP/WHNP|APRN WHN|CNM/ARNP|CNM MSN', 
        case=False, na=False), 'NP_Type_Grouped'] = 'OBGYN/Womens Health NP'

np_address.loc[np_address['NP_Type_Grouped'].isna() & np_address['Credentials'].notna() & 
    np_address['Credentials'].str.contains(
        'RN PNP|PNP', 
        case=False, na=False), 'NP_Type_Grouped'] = 'Pediatrics NP'

np_address.loc[np_address['NP_Type_Grouped'].isna() & np_address['Credentials'].notna() & 
    np_address['Credentials'].str.contains(
        'PMHNP|PHD, CAN', 
        case=False, na=False), 'NP_Type_Grouped'] = 'Psych/Mental Health NP'

np_address.loc[np_address['NP_Type_Grouped'].isna(), 'NP_Type_Grouped'] = 'Unclassified NP'

np_address

<>:4: SyntaxWarning: "\(" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\("? A raw string is also an option.
<>:4: SyntaxWarning: "\(" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\("? A raw string is also an option.
/var/folders/7q/7r1pkdyx07g6795dl8b86cs80000gp/T/ipykernel_30798/3350859588.py:4: SyntaxWarning: "\(" is an invalid escape sequence. Such sequences will not work in the future. Did you mean "\\("? A raw string is also an option.
  'ACNP|NP-ADVAN|APRN OR|A.C.N.P.|CRNFA|DCNP|AOCNP|AGPC-BC|RNFA|NP \(ACNP',


,NPI,Last_Name,First_Name,Sex,Street1,Street2,City,ZIP,Credentials,Certification_Date,Enumeration_Date,License_Number,NP_Type_Grouped,Taxonomy_Code,Taxonomy_Description,Medicare_Supplier_Type,Primary_Tax,Medicare_Specialty_Code
0,1003095977,NICHOLAS,TRACY,F,1968 PEACHTREE RD NW,PIEDMONT HO,ATLANTA,303091281,NP,NaN,10/29/2007,185022,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
1,1003104670,DODD,JOHN,M,2501 N PATTERSON ST,NaN,VALDOSTA,316021735,NP,NaN,7/18/2011,RN117801,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
2,1003128240,SARNO,JENNIFER,F,780 CANTON RD NE,SUITE 400,MARIETTA,300607241,ACNP-BC,NaN,7/3/2010,177164,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
3,1003222191,SAUCEDO,ISRAEL,M,417 W 3RD AVE,NaN,ALBANY,317011943,"APRN, FN",6/29/2022,7/9/2014,RN170526,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50
4,1003280975,SHUMAKER-KIRK,CHARVETTE,F,600 CELEBRATE LIFE PKWY,NaN,NEWNAN,302658001,NP-C,5/30/2023,11/16/2015,RN193221,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14462,1982088340,STONE,SARAH,F,3278 MITCHELL BLVD,NaN,MOODY AFB,316995529,APRN-CNP,NaN,7/16/2015,RN176410,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
14463,1982210860,VINES,KRISTI,F,10692 MEDLOCK BRIDGE RD,NaN,JOHNS CREEK,300971890,WHNP,9/16/2020,9/16/2020,RN255722,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
14464,1982679189,KINNEY,KIMBERLY,F,5354 REYNOLDS ST STE 422,NaN,SAVANNAH,314056011,"RN, NP",NaN,2/21/2006,24164385,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50
14465,1992235469,CHAPMAN,NIJAH,F,2665 N DECATUR RD STE 255,NaN,DECATUR,300336176,WHNP,NaN,6/15/2017,RN228375,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50


In [19]:
#0 means we successfully matched all previously no specialty defined NPs to a specialty
#but there are remaining NaNs: should only be rows where both Credentials and NP_Type_Grouped are missing (288 as seen in miss_cred!)
test = np_address[np_address['NP_Type_Grouped'].isna() & np_address['Credentials'].isna()]
test

,NPI,Last_Name,First_Name,Sex,Street1,Street2,City,ZIP,Credentials,Certification_Date,Enumeration_Date,License_Number,NP_Type_Grouped,Taxonomy_Code,Taxonomy_Description,Medicare_Supplier_Type,Primary_Tax,Medicare_Specialty_Code


In [20]:
#so what to do with these people???
print(test['Taxonomy_Code'].unique())
print(test['Taxonomy_Description'].unique())
print(test['Medicare_Supplier_Type'].unique())
#since all other possible identifiers are the same across all these NPs, we will just make a new category of unknown specialty :(

[]
[]
[]


In [21]:
np_address.loc[np_address['NP_Type_Grouped'].isna() & np_address['Credentials'].isna(), 'NP_Type_Grouped'] = 'Unclassified NP'
x = np_address[np_address['NP_Type_Grouped'].isna()]
print(x['Credentials'].unique())

#other noncategorizable credentials: 'RN 03380', 'RN 12088', 'RN 05389', 'EDD, RN,' (bc Doctor of Education doesn't actually do clinical care), 'MD' 

[]


In [22]:
test2 = np_address[np_address['NP_Type_Grouped']=='Unclassified NP']
test2

,NPI,Last_Name,First_Name,Sex,Street1,Street2,City,ZIP,Credentials,Certification_Date,Enumeration_Date,License_Number,NP_Type_Grouped,Taxonomy_Code,Taxonomy_Description,Medicare_Supplier_Type,Primary_Tax,Medicare_Specialty_Code
9573,1013312529,ONYIRIMBA,ESTHER,F,870 CRESTMARK DR,NaN,LITHIA SPRINGS,301222665,NaN,NaN,10/27/2014,RN205364,Unclassified NP,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
9588,1013790096,COOK,KENDRA,F,123 WEAVER RD STE B,NaN,BLAIRSVILLE,305123155,APRN/NP,8/15/2023,8/15/2023,RN236809,Unclassified NP,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
9597,1023144847,GUNN,JOYCE,F,811 HEMLOCK ST,NaN,MACON,312012144,RN 03380,NaN,2/26/2007,RN 033806 NP,Unclassified NP,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
9608,1023576170,NEWMAN,BRANDI,F,902 N 7TH ST,NaN,CORDELE,310153234,NaN,3/3/2021,3/7/2019,140369,Unclassified NP,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
9611,1023619764,ONWUBIKO,CHINWENWA,F,300 COLONIAL CENTER PKWY,NaN,ROSWELL,300764899,NaN,11/6/2020,11/6/2020,RN225814,Unclassified NP,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12095,1982239497,FU,YING,F,777 W PEACHTREE ST,NaN,NORCROSS,300711868,NaN,3/4/2020,3/4/2020,RN252505,Unclassified NP,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
12096,1982294625,BRADSHAW,KELLEY,F,1364 CLIFTON RD NE,NaN,ATLANTA,303221059,NaN,5/17/2021,1/26/2021,RN301684,Unclassified NP,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
12113,1992145155,TURNER,RACHEL,F,21 COMMERCE PKWY,NaN,ADAIRSVILLE,301032009,NaN,NaN,7/2/2013,RN202294,Unclassified NP,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
12115,1992180517,VEGA,JULIA,F,5670 PEACHTREE DUNWOODY RD STE 1100,NaN,ATLANTA,303424795,NaN,9/18/2020,7/24/2015,216789,Unclassified NP,363L00000X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50


# mapping zip (hud crosswalk)

In [14]:
crosswalk = pd.read_csv("~/Downloads/aidatalab/ZipCounty_Crosswalk.csv")
crosswalk = crosswalk[crosswalk["USPS_ZIP_PREF_STATE"] == ("GA")]
crosswalk["ZIP"] = crosswalk["ZIP"].astype(str).str[:5]
crosswalk

,ZIP,COUNTY,USPS_ZIP_PREF_CITY,USPS_ZIP_PREF_STATE,RES_RATIO,BUS_RATIO,OTH_RATIO,TOT_RATIO
14745,30002,13089,AVONDALE ESTATES,GA,1.000000,1.000000,1.00000,1.000000
14746,30003,13135,NORCROSS,GA,1.000000,1.000000,1.00000,1.000000
14747,30004,13117,ALPHARETTA,GA,0.228459,0.565332,0.59144,0.264573
14748,30004,13121,ALPHARETTA,GA,0.715821,0.427870,0.40856,0.685004
14749,30004,13057,ALPHARETTA,GA,0.055720,0.006798,0.00000,0.050424
...,...,...,...,...,...,...,...,...
21205,39886,13243,SHELLMAN,GA,0.929900,0.978723,1.00000,0.934124
21206,39886,13037,SHELLMAN,GA,0.018598,0.000000,0.00000,0.017128
21207,39897,13087,WHIGHAM,GA,0.026974,0.018519,0.00000,0.026113
21208,39897,13131,WHIGHAM,GA,0.973026,0.981481,1.00000,0.973887


In [15]:
np_address["ZIP"] = np_address["ZIP"].astype(str).str[:5]
np_address

/var/folders/7q/7r1pkdyx07g6795dl8b86cs80000gp/T/ipykernel_4082/1451419422.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  np_address["ZIP"] = np_address["ZIP"].astype(str).str[:5]


,NPI,Last_Name,First_Name,Sex,Street1,Street2,City,ZIP,Credentials,Certification_Date,Enumeration_Date,License_Number,NP_Type_Grouped,Taxonomy_Code,Taxonomy_Description,Medicare_Supplier_Type,Primary_Tax,Medicare_Specialty_Code
0,1003095977,NICHOLAS,TRACY,F,1968 PEACHTREE RD NW,PIEDMONT HO,ATLANTA,30309,NP,NaN,10/29/2007,185022,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
1,1003104670,DODD,JOHN,M,2501 N PATTERSON ST,NaN,VALDOSTA,31602,NP,NaN,7/18/2011,RN117801,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
2,1003128240,SARNO,JENNIFER,F,780 CANTON RD NE,SUITE 400,MARIETTA,30060,ACNP-BC,NaN,7/3/2010,177164,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
3,1003222191,SAUCEDO,ISRAEL,M,417 W 3RD AVE,NaN,ALBANY,31701,"APRN, FN",6/29/2022,7/9/2014,RN170526,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50
4,1003280975,SHUMAKER-KIRK,CHARVETTE,F,600 CELEBRATE LIFE PKWY,NaN,NEWNAN,30265,NP-C,5/30/2023,11/16/2015,RN193221,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14462,1982088340,STONE,SARAH,F,3278 MITCHELL BLVD,NaN,MOODY AFB,31699,APRN-CNP,NaN,7/16/2015,RN176410,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
14463,1982210860,VINES,KRISTI,F,10692 MEDLOCK BRIDGE RD,NaN,JOHNS CREEK,30097,WHNP,9/16/2020,9/16/2020,RN255722,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50
14464,1982679189,KINNEY,KIMBERLY,F,5354 REYNOLDS ST STE 422,NaN,SAVANNAH,31405,"RN, NP",NaN,2/21/2006,24164385,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50
14465,1992235469,CHAPMAN,NIJAH,F,2665 N DECATUR RD STE 255,NaN,DECATUR,30033,WHNP,NaN,6/15/2017,RN228375,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50


In [16]:
#lookup dict
zip_county_weights = (crosswalk.groupby('ZIP').apply(lambda g: (g['COUNTY'].tolist(), g['RES_RATIO'].tolist())).to_dict())

#function to assign based on res_ratio
def assign_county(ZIP):
    entry = zip_county_weights.get(ZIP)
    if not entry:
        return np.nan
        
    counties, weights = entry
    weights = np.array(weights, dtype=float)

    weights = np.where(np.isnan(weights) | (weights < 0), 0, weights)
    total = weights.sum()
    if total == 0:
        # all weights invalid → choose randomly among available counties
        return np.random.choice(counties)

    weights = np.nan_to_num(weights)
    weights = weights / weights.sum()

    return np.random.choice(counties, p=weights)

np_address['County'] = np_address['ZIP'].apply(assign_county)
np_address

/var/folders/7q/7r1pkdyx07g6795dl8b86cs80000gp/T/ipykernel_4082/4050567662.py:2: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  zip_county_weights = (crosswalk.groupby('ZIP').apply(lambda g: (g['COUNTY'].tolist(), g['RES_RATIO'].tolist())).to_dict())


,NPI,Last_Name,First_Name,Sex,Street1,Street2,City,ZIP,Credentials,Certification_Date,Enumeration_Date,License_Number,NP_Type_Grouped,Taxonomy_Code,Taxonomy_Description,Medicare_Supplier_Type,Primary_Tax,Medicare_Specialty_Code,County
0,1003095977,NICHOLAS,TRACY,F,1968 PEACHTREE RD NW,PIEDMONT HO,ATLANTA,30309,NP,NaN,10/29/2007,185022,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13121.0
1,1003104670,DODD,JOHN,M,2501 N PATTERSON ST,NaN,VALDOSTA,31602,NP,NaN,7/18/2011,RN117801,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13027.0
2,1003128240,SARNO,JENNIFER,F,780 CANTON RD NE,SUITE 400,MARIETTA,30060,ACNP-BC,NaN,7/3/2010,177164,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13067.0
3,1003222191,SAUCEDO,ISRAEL,M,417 W 3RD AVE,NaN,ALBANY,31701,"APRN, FN",6/29/2022,7/9/2014,RN170526,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50,13095.0
4,1003280975,SHUMAKER-KIRK,CHARVETTE,F,600 CELEBRATE LIFE PKWY,NaN,NEWNAN,30265,NP-C,5/30/2023,11/16/2015,RN193221,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13077.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14462,1982088340,STONE,SARAH,F,3278 MITCHELL BLVD,NaN,MOODY AFB,31699,APRN-CNP,NaN,7/16/2015,RN176410,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13185.0
14463,1982210860,VINES,KRISTI,F,10692 MEDLOCK BRIDGE RD,NaN,JOHNS CREEK,30097,WHNP,9/16/2020,9/16/2020,RN255722,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13121.0
14464,1982679189,KINNEY,KIMBERLY,F,5354 REYNOLDS ST STE 422,NaN,SAVANNAH,31405,"RN, NP",NaN,2/21/2006,24164385,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50,13051.0
14465,1992235469,CHAPMAN,NIJAH,F,2665 N DECATUR RD STE 255,NaN,DECATUR,30033,WHNP,NaN,6/15/2017,RN228375,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13089.0


In [17]:
print("Mapped:", np_address['County'].notna().sum(), "of", len(np_address))

Mapped: 14456 of 14467


In [18]:
#almost all in other states??? LOL
14467-14456

11

In [19]:
np_address["County"] = np_address["County"].astype(str).str[:5]
np_address

,NPI,Last_Name,First_Name,Sex,Street1,Street2,City,ZIP,Credentials,Certification_Date,Enumeration_Date,License_Number,NP_Type_Grouped,Taxonomy_Code,Taxonomy_Description,Medicare_Supplier_Type,Primary_Tax,Medicare_Specialty_Code,County
0,1003095977,NICHOLAS,TRACY,F,1968 PEACHTREE RD NW,PIEDMONT HO,ATLANTA,30309,NP,NaN,10/29/2007,185022,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13121
1,1003104670,DODD,JOHN,M,2501 N PATTERSON ST,NaN,VALDOSTA,31602,NP,NaN,7/18/2011,RN117801,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13027
2,1003128240,SARNO,JENNIFER,F,780 CANTON RD NE,SUITE 400,MARIETTA,30060,ACNP-BC,NaN,7/3/2010,177164,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13067
3,1003222191,SAUCEDO,ISRAEL,M,417 W 3RD AVE,NaN,ALBANY,31701,"APRN, FN",6/29/2022,7/9/2014,RN170526,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50,13095
4,1003280975,SHUMAKER-KIRK,CHARVETTE,F,600 CELEBRATE LIFE PKWY,NaN,NEWNAN,30265,NP-C,5/30/2023,11/16/2015,RN193221,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13077
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14462,1982088340,STONE,SARAH,F,3278 MITCHELL BLVD,NaN,MOODY AFB,31699,APRN-CNP,NaN,7/16/2015,RN176410,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13185
14463,1982210860,VINES,KRISTI,F,10692 MEDLOCK BRIDGE RD,NaN,JOHNS CREEK,30097,WHNP,9/16/2020,9/16/2020,RN255722,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13121
14464,1982679189,KINNEY,KIMBERLY,F,5354 REYNOLDS ST STE 422,NaN,SAVANNAH,31405,"RN, NP",NaN,2/21/2006,24164385,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50,13051
14465,1992235469,CHAPMAN,NIJAH,F,2665 N DECATUR RD STE 255,NaN,DECATUR,30033,WHNP,NaN,6/15/2017,RN228375,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13089


In [20]:
fips = pd.read_csv("~/Downloads/aidatalab/GA_FIPS.csv")
fips['FIPS Code'] = fips['FIPS Code'].astype(str)
fips

,FIPS Code,County Name
0,13001,Appling
1,13003,Atkinson
2,13005,Bacon
3,13007,Baker
4,13009,Baldwin
...,...,...
154,13313,Whitfield
155,13315,Wilcox
156,13317,Wilkes
157,13319,Wilkinson


In [21]:
fips_map = fips.set_index('FIPS Code')['County Name'].to_dict()
np_address['County'] = np_address['County'].astype(str)
np_address['County_Name'] = np_address['County'].map(fips_map)
np_address

,NPI,Last_Name,First_Name,Sex,Street1,Street2,City,ZIP,Credentials,Certification_Date,Enumeration_Date,License_Number,NP_Type_Grouped,Taxonomy_Code,Taxonomy_Description,Medicare_Supplier_Type,Primary_Tax,Medicare_Specialty_Code,County,County_Name
0,1003095977,NICHOLAS,TRACY,F,1968 PEACHTREE RD NW,PIEDMONT HO,ATLANTA,30309,NP,NaN,10/29/2007,185022,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13121,Fulton
1,1003104670,DODD,JOHN,M,2501 N PATTERSON ST,NaN,VALDOSTA,31602,NP,NaN,7/18/2011,RN117801,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13027,Brooks
2,1003128240,SARNO,JENNIFER,F,780 CANTON RD NE,SUITE 400,MARIETTA,30060,ACNP-BC,NaN,7/3/2010,177164,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13067,Cobb
3,1003222191,SAUCEDO,ISRAEL,M,417 W 3RD AVE,NaN,ALBANY,31701,"APRN, FN",6/29/2022,7/9/2014,RN170526,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50,13095,Dougherty
4,1003280975,SHUMAKER-KIRK,CHARVETTE,F,600 CELEBRATE LIFE PKWY,NaN,NEWNAN,30265,NP-C,5/30/2023,11/16/2015,RN193221,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13077,Coweta
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14462,1982088340,STONE,SARAH,F,3278 MITCHELL BLVD,NaN,MOODY AFB,31699,APRN-CNP,NaN,7/16/2015,RN176410,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13185,Lowndes
14463,1982210860,VINES,KRISTI,F,10692 MEDLOCK BRIDGE RD,NaN,JOHNS CREEK,30097,WHNP,9/16/2020,9/16/2020,RN255722,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13121,Fulton
14464,1982679189,KINNEY,KIMBERLY,F,5354 REYNOLDS ST STE 422,NaN,SAVANNAH,31405,"RN, NP",NaN,2/21/2006,24164385,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50,13051,Chatham
14465,1992235469,CHAPMAN,NIJAH,F,2665 N DECATUR RD STE 255,NaN,DECATUR,30033,WHNP,NaN,6/15/2017,RN228375,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13089,De Kalb


In [22]:
np_address.to_csv('shape_mapping.csv', index=False)

In [27]:
agg = np_address.groupby(['County_Name', 'NP_Type']).size().reset_index(name='count')
agg

,County_Name,NP_Type,count
0,Appling,Family NP,20
1,Appling,"NP, No Subspecialty Noted",6
2,Appling,Primary Care NP,2
3,Atkinson,Family NP,4
4,Atkinson,"NP, No Subspecialty Noted",1
...,...,...,...
829,Wilcox,Family NP,6
830,Wilkes,Family NP,1
831,Wilkinson,Family NP,3
832,Wilkinson,"NP, No Subspecialty Noted",1


In [ ]:
final

# mapping zip (geocoding)

In [32]:
#standardize address:

np_address['Street1'] = np_address['Street1'].fillna('')
np_address['Street2'] = np_address['Street2'].fillna('')

#convert to lower case
np_address['Street1'] = np_address['Street1'].str.upper().str.strip()
np_address['Street2'] = np_address['Street2'].str.upper().str.strip()

#punctuation
np_address['Street1'] = np_address['Street1'].str.replace(r'[^\w\s]', '', regex=True)
np_address['Street2'] = np_address['Street2'].str.replace(r'[^\w\s]', '', regex=True)

#common abbreviations
abbrev_map = {
    r'\bST\b\.?': 'STREET',
    r'\bRD\b\.?': 'ROAD',
    r'\bAVE\b\.?': 'AVENUE',
    r'\bBLVD\b\.?': 'BOULEVARD',
    r'\bDR\b\.?': 'DRIVE',
    r'\bLN\b\.?': 'LANE',
    r'\bCT\b\.?': 'COURT',
    r'\bPL\b\.?': 'PLACE',
    r'\bPKWY\b\.?': 'PARKWAY',
    r'\bHWY\b\.?': 'HIGHWAY',
    r'\bTER\b\.?': 'TERRACE',
    r'\bST\b\.?': 'SUITE',
    r'\bHO\b\.?': 'HOSPITAL',
}

for abbrev, full in abbrev_map.items():
    np_address['Street1'] = np_address['Street1'].str.replace(abbrev, full, regex=True)
    np_address['Street2'] = np_address['Street2'].str.replace(abbrev, full, regex=True)

#all zip codes to first 5 digits
np_address["ZIP"] = np_address["ZIP"].astype(str).str[:5]
np_address

np_address

,NPI,Last_Name,First_Name,Sex,Street1,Street2,City,ZIP,Credentials,Certification_Date,Enumeration_Date,License_Number,NP_Type_Grouped,Taxonomy_Code,Taxonomy_Description,Medicare_Supplier_Type,Primary_Tax,Medicare_Specialty_Code,County,County_Name
0,1003095977,NICHOLAS,TRACY,F,1968 PEACHTREE ROAD NW,PIEDMONT HOSPITAL,ATLANTA,30309,NP,NaN,10/29/2007,185022,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13121,Fulton
1,1003104670,DODD,JOHN,M,2501 N PATTERSON SUITE,,VALDOSTA,31602,NP,NaN,7/18/2011,RN117801,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13185,Lowndes
2,1003128240,SARNO,JENNIFER,F,780 CANTON ROAD NE,SUITE 400,MARIETTA,30060,ACNP-BC,NaN,7/3/2010,177164,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13067,Cobb
3,1003222191,SAUCEDO,ISRAEL,M,417 W 3RD AVENUE,,ALBANY,31701,"APRN, FN",6/29/2022,7/9/2014,RN170526,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50,13095,Dougherty
4,1003280975,SHUMAKER-KIRK,CHARVETTE,F,600 CELEBRATE LIFE PARKWAY,,NEWNAN,30265,NP-C,5/30/2023,11/16/2015,RN193221,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13077,Coweta
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14462,1982088340,STONE,SARAH,F,3278 MITCHELL BOULEVARD,,MOODY AFB,31699,APRN-CNP,NaN,7/16/2015,RN176410,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13185,Lowndes
14463,1982210860,VINES,KRISTI,F,10692 MEDLOCK BRIDGE ROAD,,JOHNS CREEK,30097,WHNP,9/16/2020,9/16/2020,RN255722,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13135,Gwinnett
14464,1982679189,KINNEY,KIMBERLY,F,5354 REYNOLDS SUITE STE 422,,SAVANNAH,31405,"RN, NP",NaN,2/21/2006,24164385,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50,13051,Chatham
14465,1992235469,CHAPMAN,NIJAH,F,2665 N DECATUR ROAD STE 255,,DECATUR,30033,WHNP,NaN,6/15/2017,RN228375,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13089,De Kalb


In [33]:
#combine to one address line
np_address['full_address'] = (
    np_address['Street1'] + ' ' +
    np_address['Street2'] + ', ' +
    np_address['City'] + ', ' +
    'GA'
)

np_address

,NPI,Last_Name,First_Name,Sex,Street1,Street2,City,ZIP,Credentials,Certification_Date,Enumeration_Date,License_Number,NP_Type_Grouped,Taxonomy_Code,Taxonomy_Description,Medicare_Supplier_Type,Primary_Tax,Medicare_Specialty_Code,County,County_Name,full_address
0,1003095977,NICHOLAS,TRACY,F,1968 PEACHTREE ROAD NW,PIEDMONT HOSPITAL,ATLANTA,30309,NP,NaN,10/29/2007,185022,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13121,Fulton,"1968 PEACHTREE ROAD NW PIEDMONT HOSPITAL, ATLA..."
1,1003104670,DODD,JOHN,M,2501 N PATTERSON SUITE,,VALDOSTA,31602,NP,NaN,7/18/2011,RN117801,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13185,Lowndes,"2501 N PATTERSON SUITE , VALDOSTA, GA"
2,1003128240,SARNO,JENNIFER,F,780 CANTON ROAD NE,SUITE 400,MARIETTA,30060,ACNP-BC,NaN,7/3/2010,177164,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13067,Cobb,"780 CANTON ROAD NE SUITE 400, MARIETTA, GA"
3,1003222191,SAUCEDO,ISRAEL,M,417 W 3RD AVENUE,,ALBANY,31701,"APRN, FN",6/29/2022,7/9/2014,RN170526,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50,13095,Dougherty,"417 W 3RD AVENUE , ALBANY, GA"
4,1003280975,SHUMAKER-KIRK,CHARVETTE,F,600 CELEBRATE LIFE PARKWAY,,NEWNAN,30265,NP-C,5/30/2023,11/16/2015,RN193221,Acute Care NP,363LA2100X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13077,Coweta,"600 CELEBRATE LIFE PARKWAY , NEWNAN, GA"
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14462,1982088340,STONE,SARAH,F,3278 MITCHELL BOULEVARD,,MOODY AFB,31699,APRN-CNP,NaN,7/16/2015,RN176410,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13185,Lowndes,"3278 MITCHELL BOULEVARD , MOODY AFB, GA"
14463,1982210860,VINES,KRISTI,F,10692 MEDLOCK BRIDGE ROAD,,JOHNS CREEK,30097,WHNP,9/16/2020,9/16/2020,RN255722,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13135,Gwinnett,"10692 MEDLOCK BRIDGE ROAD , JOHNS CREEK, GA"
14464,1982679189,KINNEY,KIMBERLY,F,5354 REYNOLDS SUITE STE 422,,SAVANNAH,31405,"RN, NP",NaN,2/21/2006,24164385,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,N,50,13051,Chatham,"5354 REYNOLDS SUITE STE 422 , SAVANNAH, GA"
14465,1992235469,CHAPMAN,NIJAH,F,2665 N DECATUR ROAD STE 255,,DECATUR,30033,WHNP,NaN,6/15/2017,RN228375,OBGYN/Womens Health NP,363LW0102X,Physician Assistants & Advanced Practice Nursi...,Nurse Practitioner,Y,50,13089,De Kalb,"2665 N DECATUR ROAD STE 255 , DECATUR, GA"
